Lib/test/test_typing.py

```diff
@@ -255,6 +256,449 @@ def test_no_bivariant(self):
+        self.assertEqual(
+            a.__annotations__,
+            {'args': Unpack[tuple[int, Unpack[Ts]]]},
+        )
+
+        def b(*args: Unpack[tuple[Unpack[Ts], int]]): pass
+        self.assertEqual(
+            b.__annotations__,
+            {'args': Unpack[tuple[Unpack[Ts], int]]},
+        )
+
+        def c(*args: Unpack[tuple[str, Unpack[Ts], int]]): pass
+        self.assertEqual(
+            c.__annotations__,
+            {'args': Unpack[tuple[str, Unpack[Ts], int]]},
+        )
+
+        def d(*args: Unpack[tuple[int, bool, Unpack[Ts], float, str]]): pass
+        self.assertEqual(
+            d.__annotations__,
+            {'args': Unpack[tuple[int, bool, Unpack[Ts], float, str]]},
+        )
+
+        # Unpacking using `Unpack`, `Tuple` type from typing.py
+
+        def e(*args: Unpack[Tuple[int, Unpack[Ts]]]): pass
+        self.assertEqual(
+            e.__annotations__,
+            {'args': Unpack[Tuple[int, Unpack[Ts]]]},
+        )
+
+        def f(*args: Unpack[Tuple[Unpack[Ts], int]]): pass
+        self.assertEqual(
+            f.__annotations__,
+            {'args': Unpack[Tuple[Unpack[Ts], int]]},
+        )
+
+        def g(*args: Unpack[Tuple[str, Unpack[Ts], int]]): pass
+        self.assertEqual(
+            g.__annotations__,
+            {'args': Unpack[Tuple[str, Unpack[Ts], int]]},
+        )
+
+        def h(*args: Unpack[Tuple[int, bool, Unpack[Ts], float, str]]): pass
+        self.assertEqual(
+            h.__annotations__,
+            {'args': Unpack[Tuple[int, bool, Unpack[Ts], float, str]]},
+        )
+
+    def test_variadic_class_same_args_results_in_equalty(self):
+        Ts = TypeVarTuple('Ts')
+        class C(Generic[Unpack[Ts]]): pass
+
+        self.assertEqual(C[int], C[int])
+
+        Ts1 = TypeVarTuple('Ts1')
+        Ts2 = TypeVarTuple('Ts2')
+        self.assertEqual(
+            C[Unpack[Ts1]],
+            C[Unpack[Ts1]],
+        )
+        self.assertEqual(
+            C[Unpack[Ts1], Unpack[Ts2]],
+            C[Unpack[Ts1], Unpack[Ts2]],
+        )
+        self.assertEqual(
+            C[int, Unpack[Ts1], Unpack[Ts2]],
+            C[int, Unpack[Ts1], Unpack[Ts2]],
+        )
+
+    def test_variadic_class_arg_ordering_matters(self):
+        Ts = TypeVarTuple('Ts')
+        class C(Generic[Unpack[Ts]]): pass
+
+        self.assertNotEqual(
+            C[int, str],
+            C[str, int],
+        )
+
+        Ts1 = TypeVarTuple('Ts1')
+        Ts2 = TypeVarTuple('Ts2')
+        self.assertNotEqual(
+            C[Unpack[Ts1], Unpack[Ts2]],
+            C[Unpack[Ts2], Unpack[Ts1]],
+        )
+
+    def test_variadic_class_arg_typevartuple_identity_matters(self):
+        Ts = TypeVarTuple('Ts')
+        class C(Generic[Unpack[Ts]]): pass
+        Ts1 = TypeVarTuple('Ts1')
+        Ts2 = TypeVarTuple('Ts2')
+        self.assertNotEqual(C[Unpack[Ts1]], C[Unpack[Ts2]])
+
+    def test_typevar_substitution(self):
```

Comment:
Oh yikes, good point.

After an exhausting couple of hours, I've just added a first attempt at implementing support for this, but I still haven't covered the case where one of the type arguments is an `Unpack[Ts]` - I have a feeling that's going to be even more complicated...

> Things get a bit complicated with `tuple[int, ...]`. For example, when the expected types are `(T1, T2)`, we should have `T1=int, T2=int`

Hmm, should we? While thinking it through for the implementation I just added, I was thinking cases like this actually _shouldn't_ work, because `tuple[int, ...]` could contain _zero_ types. Wouldn't accepting this case be unsound?

Reply:
@mrahtz

> > Things get a bit complicated with tuple[int, ...]. For example, when the expected types are (T1, T2), we should have T1=int, T2=int
> Hmm, should we? While thinking it through for the implementation I just added, I was thinking cases like this actually shouldn't work, because tuple[int, ...] could contain zero types. Wouldn't accepting this case be unsound?

We'd discussed this precise case in https://github.com/python/peps/pull/2162#discussion_r762324527. tl;dr: `tuple[int, ...]` is by its nature unsound - `xs: tuple[int, ...]; y1: int = xs[0]; y2: int = xs[1]` is not a type error. It's the same idea with `xs: list[int]; y1: int = xs[0]`. Yes, both snippets can blow up at runtime if the tuple or list happen to be empty, but that's the pragmatic tradeoff we've already chosen in the language to make these types usable.

> I was thinking cases like this actually shouldn't work, because tuple[int, ...] could contain zero types.

What you're describing is treating `tuple[int, ...]` as equivalent to `tuple[()]` - a zero-length tuple. That would not serve the purpose of an *arbitrary*-length tuple. Lmk if that makes sense.

--------

> Things get a bit complicated with tuple[int, ...]. For example, when the expected types are (T1, T2), we should have T1=int, T2=int. When the expected types are (T1, *Ts, T2) and the type argument is (int, Unpack[tuple[str, ...]]), we should have T1=int, Ts=tuple[str, ...], T2=str.
> It would be great to avoid implementing all this logic, but it looks like GenericAlias demands this behavior at runtime so that people can do MyAlias[<...>] == <the instantiated type>, so we may not have a choice.

This could be a follow-up PR if you feel this PR has been dragging on for a while :)

Also curious about opinions from @JelleZijlstra  et al about what to do here at runtime. How much do we care about instantiating GenericAliases accurately at runtime? For example, who depends on the runtime behavior of `MyAlias = Array[T1, T2]; x: MyAlias[int, str]`? Type checkers handle this correctly. Do we need this to be accurately instantiated at runtime? Things get complicated for aliases involving variadic tuples.

Reply:
> This could be a follow-up PR if you feel this PR has been dragging on for a while :)

Mm, great idea. For now I've excised all this logic. I'll submit a proper solution to this in a follow-up PR.